include/pgduckdb/types/pgduckdb_enum.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "pgduckdb/duckdb_vendor/enum_type_info_templated.hpp"
+
+namespace pgduckdb {
+
+using duckdb::EnumTypeInfo;
+using duckdb::EnumTypeInfoTemplated;
+using duckdb::ExtraTypeInfo;
+using duckdb::InternalException;
+using duckdb::LogicalType;
+using duckdb::LogicalTypeId;
+using duckdb::make_shared_ptr;
+using duckdb::PhysicalType;
+using duckdb::shared_ptr;
+using duckdb::Vector;
+
+template <class T>
+class PGDuckDBEnumTypeInfo : public EnumTypeInfoTemplated<T> {
+public:
+	PGDuckDBEnumTypeInfo(Vector &values_insert_order_p, idx_t dict_size_p, Vector &oid_vec)
+	    : EnumTypeInfoTemplated<T>(values_insert_order_p, dict_size_p), oid_vec(oid_vec) {
+	}
+
+public:
+	const Vector &
+	GetMemberOids() const {
+		return oid_vec;
+	}
+
+private:
+	Vector oid_vec;
```

Comment:
Instead of storing these oids - can't we do a reverse lookup of the oid based on the string name when converting back? That way we don't need to modify the DuckDB enum type.

Reply:
I don't think there's really a way around creating a custom DuckDB enum type. I agree in theory we could get away with not storing these OIDs, but to be able to look them up we'd need to know the oid (or name) of the enum type itself. The only way I can think of to find that OID is by storing it in the type DuckDB enum type again. Which brings us back to the point where we need to modify the DuckDB enum type, and if we have to do that we might as well store the oid_vec in there for efficience too.

To clarify postgres behaviour: enums even if they have the exact same members, are not the same. And the members have different oids based on the specific enum they are part of: 
```
> create type abc as enum ('a', 'b', 'c');
> create type abc2 as enum ('a', 'b', 'c');
> select 'a'::abc = 'a'::abc2;
ERROR:  42883: operator does not exist: abc = abc2
LINE 1: select 'a'::abc = 'a'::abc2;
> select * from pg_enum;
  oid   │ enumtypid │ enumsortorder │ enumlabel
────────┼───────────┼───────────────┼───────────
 101258 │    101256 │             1 │ a
 101260 │    101256 │             2 │ b
 101262 │    101256 │             3 │ c
 101266 │    101264 │             1 │ a
 101268 │    101264 │             2 │ b
 101270 │    101264 │             3 │ c
```

Reply:
One way we might be able to get away with this is by storing the extra data in the `Vector values_insert_order;`
Vector's don't have a size, that's why the EnumTypeInfo also has a accompanying `idx_t dict_size;`, I could allocate extra data in the Vector to store this data, if the `dict_size` stays the same it should never be looked at

(This will probably fail in debug mode because of (de)serialization roundtrip verification)

Reply:
```patch
diff --git a/include/pgduckdb/types/pgduckdb_enum.hpp b/include/pgduckdb/types/pgduckdb_enum.hpp
index 7cc7e1c..f141b70 100644
--- a/include/pgduckdb/types/pgduckdb_enum.hpp
+++ b/include/pgduckdb/types/pgduckdb_enum.hpp
@@ -46,12 +46,12 @@ private:
 };
 
 struct PGDuckDBEnum {
-	static LogicalType CreateEnumType(Vector &ordered_data, idx_t size, Vector &enum_member_oids);
+	//static LogicalType CreateEnumType(Vector &ordered_data, idx_t size, Vector &enum_member_oids);
 	static idx_t GetDuckDBEnumPosition(duckdb::Value &val);
 	static int GetEnumPosition(Datum enum_member_oid);
 	static bool IsEnumType(Oid type_oid);
 	static Oid GetEnumTypeOid(const Vector &oids);
-	static const Vector &GetMemberOids(const duckdb::LogicalType &type);
+	static Vector GetMemberOids(const duckdb::LogicalType &type);
 };
 
 } // namespace pgduckdb
diff --git a/src/pgduckdb_types.cpp b/src/pgduckdb_types.cpp
index e7609ac..e73f08e 100644
--- a/src/pgduckdb_types.cpp
+++ b/src/pgduckdb_types.cpp
@@ -327,7 +327,7 @@ ConvertDuckToPostgresArray(TupleTableSlot *slot, duckdb::Value &value, idx_t col
 void
 ConvertDuckToPostgresEnumValue(TupleTableSlot *slot, duckdb::Value &val, idx_t col) {
 	auto position = PGDuckDBEnum::GetDuckDBEnumPosition(val);
-	auto &enum_oids = PGDuckDBEnum::GetMemberOids(val.type());
+	auto enum_oids = PGDuckDBEnum::GetMemberOids(val.type());
 	auto enum_value_oid = duckdb::FlatVector::GetData<Oid>(enum_oids)[position];
 	slot->tts_values[col] = ObjectIdGetDatum(enum_value_oid);
 }
@@ -544,10 +544,13 @@ ConvertPostgresEnumToDuckEnum(Oid enum_type_oid) {
 
 	std::sort(enum_members.begin(), enum_members.end(), sort_order_cmp);
 
-	auto duck_enum_vec = duckdb::Vector(duckdb::LogicalType::VARCHAR, enum_members.size());
-	auto enum_oid_vec = duckdb::Vector(duckdb::LogicalType::UINTEGER, enum_members.size());
+	idx_t allocation_size = enum_members.size();
+	allocation_size += enum_members.size() / 4;
+	allocation_size += (enum_members.size() % 4) != 0;
+
+	auto duck_enum_vec = duckdb::Vector(duckdb::LogicalType::VARCHAR, allocation_size);
 	auto enum_vec_data = duckdb::FlatVector::GetData<duckdb::string_t>(duck_enum_vec);
-	auto enum_oid_data = duckdb::FlatVector::GetData<uint32_t>(enum_oid_vec);
+	auto enum_oid_data = (uint32_t*)(enum_vec_data + enum_members.size());
 	for (idx_t i = 0; i < enum_members.size(); i++) {
 		auto &member = enum_members[i];
 		auto enum_data = (Form_pg_enum)GETSTRUCT(member);
@@ -556,7 +559,7 @@ ConvertPostgresEnumToDuckEnum(Oid enum_type_oid) {
 	}
 
 	PostgresFunctionGuard(ReleaseCatCacheList, list);
-	return PGDuckDBEnum::CreateEnumType(duck_enum_vec, enum_members.size(), enum_oid_vec);
+	return EnumTypeInfo::CreateType(duck_enum_vec, enum_members.size());
 }
 
 duckdb::LogicalType
@@ -683,7 +686,7 @@ GetPostgresDuckDBType(duckdb::LogicalType type) {
 		}
 	}
 	case duckdb::LogicalTypeId::ENUM: {
-		auto &member_oids = PGDuckDBEnum::GetMemberOids(type);
+		auto member_oids = PGDuckDBEnum::GetMemberOids(type);
 		return PGDuckDBEnum::GetEnumTypeOid(member_oids);
 	}
 	default: {
diff --git a/src/types/pgduckdb_enum.cpp b/src/types/pgduckdb_enum.cpp
index 9b85ef5..a98aa66 100644
--- a/src/types/pgduckdb_enum.cpp
+++ b/src/types/pgduckdb_enum.cpp
@@ -5,27 +5,27 @@ namespace pgduckdb {
 
 // This is copied from 'duckdb::EnumTypeInfo::CreateType' with our additional metadata added
 
-LogicalType
-PGDuckDBEnum::CreateEnumType(Vector &ordered_data, idx_t size, Vector &enum_member_oids) {
-	// Generate EnumTypeInfo
-	shared_ptr<ExtraTypeInfo> info;
-	auto enum_internal_type = EnumTypeInfo::DictType(size);
-	switch (enum_internal_type) {
-	case PhysicalType::UINT8:
-		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint8_t>>(ordered_data, size, enum_member_oids);
-		break;
-	case PhysicalType::UINT16:
-		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint16_t>>(ordered_data, size, enum_member_oids);
-		break;
-	case PhysicalType::UINT32:
-		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint32_t>>(ordered_data, size, enum_member_oids);
-		break;
-	default:
-		throw InternalException("Invalid Physical Type for ENUMs");
-	}
-	// Generate Actual Enum Type
-	return LogicalType(LogicalTypeId::ENUM, info);
-}
+//LogicalType
+//PGDuckDBEnum::CreateEnumType(Vector &ordered_data, idx_t size, Vector &enum_member_oids) {
+//	// Generate EnumTypeInfo
+//	shared_ptr<ExtraTypeInfo> info;
+//	auto enum_internal_type = EnumTypeInfo::DictType(size);
+//	switch (enum_internal_type) {
+//	case PhysicalType::UINT8:
+//		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint8_t>>(ordered_data, size, enum_member_oids);
+//		break;
+//	case PhysicalType::UINT16:
+//		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint16_t>>(ordered_data, size, enum_member_oids);
+//		break;
+//	case PhysicalType::UINT32:
+//		info = make_shared_ptr<PGDuckDBEnumTypeInfo<uint32_t>>(ordered_data, size, enum_member_oids);
+//		break;
+//	default:
+//		throw InternalException("Invalid Physical Type for ENUMs");
+//	}
+//	// Generate Actual Enum Type
+//	return LogicalType(LogicalTypeId::ENUM, info);
+//}
 
 idx_t
 PGDuckDBEnum::GetDuckDBEnumPosition(duckdb::Value &val) {
@@ -103,20 +103,14 @@ PGDuckDBEnum::GetEnumTypeOid(const Vector &oids) {
 	return result;
 }
 
-const Vector &
-PGDuckDBEnum::GetMemberOids(const duckdb::LogicalType &type) {
+Vector PGDuckDBEnum::GetMemberOids(const duckdb::LogicalType &type) {
 	auto type_info = type.AuxInfo();
 	auto &enum_type_info = type_info->Cast<duckdb::EnumTypeInfo>();
-	switch (enum_type_info.DictType(enum_type_info.GetDictSize())) {
-	case PhysicalType::UINT8:
-		return enum_type_info.Cast<PGDuckDBEnumTypeInfo<uint8_t>>().GetMemberOids();
-	case PhysicalType::UINT16:
-		return enum_type_info.Cast<PGDuckDBEnumTypeInfo<uint16_t>>().GetMemberOids();
-	case PhysicalType::UINT32:
-		return enum_type_info.Cast<PGDuckDBEnumTypeInfo<uint32_t>>().GetMemberOids();
-	default:
-		throw InternalException("Invalid Physical Type for ENUMs");
-	}
+	auto &enum_sort_order = enum_type_info.GetValuesInsertOrder();
+	idx_t dict_size = enum_type_info.GetDictSize();
+
+	auto vector_data = duckdb::FlatVector::GetData<duckdb::string_t>(enum_sort_order);
+	return Vector(duckdb::LogicalType::UINTEGER, (duckdb::data_ptr_t)(vector_data + dict_size));
 }
 
 } // namespace pgduckdb
diff --git a/third_party/duckdb b/third_party/duckdb
--- a/third_party/duckdb
+++ b/third_party/duckdb
@@ -1 +1 @@
-Subproject commit fa5c2fe15f3da5f32397b009196c0895fce60820
+Subproject commit fa5c2fe15f3da5f32397b009196c0895fce60820-dirty
```

This works 🤢

Reply:
And it's pushed

Reply:
Yea I talked about it with Mark, there isn't a good solution and indeed all of them have a certain ick to them

Reply:
Hmm now trying to make `select a > 'sad' from tbl;` work, which involves storing the LogicalType in the TypeCatalogEntry, this invokes `ExtraTypeInfo::Copy` and breaks this "solution"

```c++
shared_ptr<ExtraTypeInfo> EnumTypeInfo::Copy() const {
	Vector values_insert_order_copy(LogicalType::VARCHAR, false, false, 0);
	values_insert_order_copy.Reference(values_insert_order);
	return make_shared_ptr<EnumTypeInfo>(values_insert_order_copy, dict_size);
}
```

Reply:
The actual issue was this:
```patch
diff --git a/src/pgduckdb_filter.cpp b/src/pgduckdb_filter.cpp
index 081271d..11c9bb2 100644
--- a/src/pgduckdb_filter.cpp
+++ b/src/pgduckdb_filter.cpp
@@ -71,8 +71,7 @@ FilterOperationSwitch(Datum &value, duckdb::Value &constant, Oid type_oid) {
 	case VARCHAROID:
 		return StringFilterOperation<OP>(value, constant);
 	default:
-		throw duckdb::InvalidTypeException(
-		    duckdb::string("(DuckDB/FilterOperationSwitch) Unsupported duckdb type: %d", type_oid));
+		throw duckdb::InvalidTypeException(duckdb::StringUtil::Format("(DuckDB/FilterOperationSwitch) Unsupported duckdb type: %d", type_oid));
 	}
 }
```

Reply:
This invoked this `std::string` constructor:
```
basic_string( const basic_string& other, size_type pos,
              const Allocator& alloc = Allocator() );
```

Which flooded my output with *every* string literal that duckdb contains 🙃